# LAB: "Rice GWAS"
## Rice GWAS Summary:
- goal: 
  - examine and summarize some rice trait data
  - perform GWAS for a trait
  - find canidate genes underlying a GWAS peak
- Install `statgenGWAS`
- Get data w/ pop assignments from fastStrucutre
- Choose a trait to work on like `seed.length`
  - plot single histogram for all the data, separate histograms for each pop, and a boxplot based on pop assignment
- Perform ANOVA to see if trait varies b/w pop
- GWAS
  - used to look for significant assoc b/w allele sttes at SNPs asnd phenotypic traits
  - tests each SNP in turn
- `statGWAS` needs 4 data frames
  - gt data - contains gt dta at each snp for each ind
  - gt map - contains info of location of each SNP
  - pheno df - pheno info for each ind
  - kinship matrix - pairwise genetic relat4edness b/w ind
  - need to make minor changes to get it to work
  - also need to put it all in one object
- RUN GWAS
  - w/o correction there are a lot of "significant SNPs"
  - Use different population corrections like PCA, Kinship matrix
- Get significant SNPs
  - sort to find the most significant SNP, and find it on the rice genome
  - 
