---
title: "Assignment 5"
output: 
  html_document:
    df_print: paged
---
__Name:__ Ricardo Pineda

__Student ID:__ 917486212

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, max.print=100)
```

```{r Library Calls}

library(tidyverse)
```

## Lab Summary
- Have data with SNPs 
  - Individual | gt @ marker
- Make a subset of this data set to reduce compute time when it's limited like in the class
  - take random rows of the SNP data
- Turn gt -> numeric
- remove NAs
  -quick way by just using the average @ SNP
- Compute PCs
  - get data
    - PCs, sdev of PCs, rotation, center and scale
    - each column is PC and each row is sample
- Find proportion of variance explained by each PC by using sdev^2
- plot the major PCs 
- Adding pheno via join
  - use join by having some shared bit (ID)
- Assign pop using faststructure
  - need fam file with descriptions of ind in study
  - also bim with info for alleles
- get results wth proportion of what amount of a sample's genome came from a certain pop.
- wide -> long to plot 
  - wide is 1 row = 1 ind
  - long is 1 row = 1 observation



## Assignment 5: Rice SNPs

Remember to include the relevant code from the lab page in this file so that this file will knit.

```{r Set Up}
Sys.setenv(VROOM_CONNECTION_SIZE="500000") # needed because the lines in this file are _extremely_long.
data.geno <- read_csv("../input/Rice_44K_genotypes.csv.gz",
                      na=c("NA","00")) #this tells R that missing data is denoted as "NA" or "00"
```
```{r data.geno structure and fix column names}
#head(data.geno[,1:10])
#summary(data.geno[,1:10]) #summarize the first 10 columns
data.geno <- data.geno %>% rename(ID=...1)
head(data.geno[,1:10])
summary(data.geno[,1:10])
```

```{r Learning sample}
#sample(x = 1:10, size = 5) # x is data size us # samples
#m2 <- matrix(1:100, nrow = 10)
# makes 10x10 matrix going 1-10/column to 100
# m2[sample(1:nrow(m2),3),] m2 randomy samples 3 rows of original m2
```

**Exercise 0:** Be sure that your final knitted file has good formatting.  Make sure that you are using informative variable names.

**Exercise 1** Create a data subset that contains a random sample of 10000 SNPs from the full data set.  Place the smaller data set in an object called `data.geno.10000`.  __Very important: you want to keep the first column, the one with the variety IDs, and you want it to be the first column in `data.geno.10000`.  AND You do not want it to show up randomly later on in the data set.__  Think about how to achieve this. _Hint: `[ ]` will help_

```{r Exercise 1}
set.seed(0421)

#Probably a better way to do this
data.geno.10000 <- cbind(data.geno[,1])
data.geno.10000 <- mutate(data.geno.10000, data.geno[,sample(2:ncol(data.geno),10000)])

# Can also do it this way
#data.geno.10000 <- cbind(data.geno[,1]) %>%
#  mutate(data.geno.10000, data.geno[,sample(2:ncol(data.geno),10000)])

dim(data.geno.10000)
colnames(data.geno.10000) %>% str_which("ID")
```

**Exercise 2:** plot the variance explained by the first 10 PCs.

```{r Exercise 2 Set Up 1}
geno.numeric <- data.geno.10000[,-1] %>% #remove ID
  lapply(factor) %>% #Char -> Factor
  as.data.frame() %>% #reformat?
  data.matrix() # as num

#head(geno.numeric[,1:10])

geno.numeric.fill <-
  apply(geno.numeric, 2, function(x) {
    x[is.na(x)] <- mean(x, na.rm = T)
    x})

geno.pca <- prcomp(geno.numeric.fill,
                   rank.= 10)

#str(geno.pca)
# x =  PC, sdev = SD of PC, rotation = loadings?, center & scale = center and scaling factors

#head(geno.pca$x)[,1:5] #PC themselves
#head(geno.pca$rotation)[,1:5] #SNP contribution
#head(geno.pca$sdev) #SD explained by each PC
```

```{r Exercise 2 Set Up 2}
pcvar <- geno.pca$sdev^2 #get var from SD
pcvar.pct <- tibble(pctvar =  pcvar /sum(pcvar) * 100,
                    PC = 1:length(pcvar))
```

```{r Exercise 2}
ggplot(data = pcvar.pct[1:10,], aes(x = PC, y  = pctvar)) +
  geom_col() +
  ggtitle("Percent Variance Explained")
```

**Exercise 3:** Make 2 scatter plots, the first of PC1 vs PC2, and second PC2 vs PC3 (keep PC2 on the y-axis for both plots). Is there any evidence for populations structure (different sub populations)?  If so, how many sub populations do you think the PCA plot reveals?  What do you make of the individuals that are between the major groups?

```{r Exercise 3 Set Up}
PCs <- as_tibble(geno.pca$x) %>%
  mutate(ID =  data.geno.10000$ID) %>%
  select(ID, everything())
head(PCs)
```

```{r Exercise 3}

ggplot(data = PCs, aes(x = PC1, y  = PC2)) +
  geom_point() +
  ggtitle("PC1 vs PC2")

ggplot(data = PCs, aes(x = PC3, y  = PC2)) +
  geom_point() +
  ggtitle("PC3 vs PC2")
```


- It appears that there is evidence of population structure from the clusters present in the plot. I think the PCA plots show 3 or 4 sub-populations and the individuals that are between the major groups may have have a mixed genome in between the genome that characterizes each population.

```{r MDS compared to PCA}
genDist <- as.matrix(dist(geno.numeric))
geno.mds <- as_tibble(cmdscale(genDist))
geno.mds$ID <- data.geno.10000$ID
#head(geno.mds)

ggplot(data = geno.mds, aes(x = V1, y = V2)) +
  geom_point() +
  ggtitle("MDS Plot")
```

**Exercise 4:** 
* Use the `read_csv()` `head()` and `summary()` (or `skim()`...look it up) functions that you learned earlier to import and look at this file.  Import the file into an object called "data.pheno".
* Use a `join` function to merge the PC  genotype data (in the object `PCs`) with the phenotype data into a new object called `data.pheno.pca`. Use summary and head to look at the new object and make sure that it is as you expect.  
* It (`data.pheno.pca`) should have 413 rows and 49 columns.
* Include your code in the .Rmd

```{r Exercise 4}

data.pheno <- read_csv("../input/RiceDiversity.44K.MSU6.Phenotypes.csv")
#str(data.pheno)
#str(PCs)
data.pheno.pca <- left_join(data.pheno, PCs, by = c("NSFTVID"="ID"))
#str(data.pheno.pca)
dim(data.pheno.pca)
```


**Exercise 5:** Prepare three different PCA plots to explore if subgroups vary by 1) Amylose content; 2) Pericarp color; 3) Region.  That is make a scatter plot of PC1 vs PC2 and color the points by the above characteristics.  Do any of these seem to be associated with the different population groups? Briefly discuss.  (optionally repeat the plots plotting PC2 vs PC3)

```{r Exercise 5}
ggplot(data = data.pheno.pca, aes(x = PC1, y  = PC2, color = `Amylose content`)) +
  geom_point() +
  ggtitle("PC1 and PC2 - Amylose Content")

ggplot(data = data.pheno.pca, aes(x = PC1, y  = PC2, color = `Pericarp color`)) +
  geom_point() +
  ggtitle("PC1 and PC2 - Pericarp")

ggplot(data = data.pheno.pca, aes(x = PC1, y  = PC2, color = Region)) +
  geom_point() +
  ggtitle("PC1 and PC2 - Region") +
  scale_color_brewer(type="qual", palette = "Set1")
  #Adding the scale color brewer doesn't plot the NA's
```

```{r OPTIONAL Exercise 5}
ggplot(data = data.pheno.pca, aes(x = PC2, y  = PC3, color = `Amylose content`)) +
  geom_point() +
  ggtitle("PC2 and PC3")

ggplot(data = data.pheno.pca, aes(x = PC2, y  = PC3, color = `Pericarp color`)) +
  geom_point() +
  ggtitle("PC2 and PC3")

ggplot(data = data.pheno.pca, aes(x = PC2, y  = PC3, color = Region)) +
  geom_point() +
  ggtitle("PC2 and PC3") +
  scale_color_brewer(type="qual", palette = "Set1")
```

- Based on the PC1 vs PC2 plots, it appears that there is some association for these traits b/w the clusters. 
    -For the amylose plot, the clusters on the right have higher amylose content compared the cluster on the left, which mostly has low amylose content.
    - For pericarp color, the populations on the top are associated with a light color, while the population on the bottom has some association with a dark pericarp.
    - For region, there appears to be less association as one population is very mixed. The other populations are either mosty S Asian or E Asian.

**Exercise 6:** First, use a join function to combine the PCA data (in object `PCs`) with the population assignments (in `fs_results`) and place the result in `geno.pca.pop`  Then re plot the PCA data, but include the population assignment in an informative way.  How do the populations assignments relate to the PCA plot?


```{r Exercise 6 Set Up FAM}
#Create a new Matrix to hold reformatted data
data.geno.10000.fs <- matrix("",nrow=nrow(data.geno.10000)*2,ncol=ncol(data.geno.10000)-1+6)

#for each row of genotypes, create 2 rows, one with the first allele and one with the second allele.
#additionally, fill the first 6 columns with the strain ID (could also be blank but we need something there)
for (i in 1:nrow(data.geno.10000)) {
  data.geno.10000.fs[(i-1)*2+1,1:6] <- data.geno.10000[[i,1]]
  data.geno.10000.fs[(i-1)*2+2,1:6] <- data.geno.10000[[i,1]]
  data.geno.10000.fs[(i-1)*2+1,-1:-6] <- substr(data.geno.10000[i,-1],1,1)
  data.geno.10000.fs[(i-1)*2+2,-1:-6] <- substr(data.geno.10000[i,-1],2,2)
}

data.geno.10000.fs[is.na(data.geno.10000.fs)] <- -9 # fastStructure's code for missing data

dim(data.geno.10000.fs)

#take a look
data.geno.10000.fs[1:10,1:10]

write.table(data.geno.10000.fs,file="../output/rice.data.fastStructure.input.str", col.names = FALSE, row.names = FALSE, quote = FALSE)
```

```{r Exercise 6 Set Up 2 FAM}
fam <- tibble(
  FID=data.geno.10000$ID,
  IID=data.geno.10000$ID,
  PID=0,
  MID=0,
  Sex=0,
  Ptype=-9)

head(fam)
write.table(fam,file="../output/rice.data.fastStructure.input.fam",col.names = FALSE, row.names = FALSE, quote = FALSE)
```

```{r Exercise 6 Set Up BIM}
bim <- data.geno.10000.fs[,-1:-6]

colnames(bim) <- colnames(data.geno.10000)[-1]

bim[bim=="-9"] <- NA

bim <- apply(bim,2,function(x) unique(na.omit(x))) 

bim[,1:5]

bim <- t(bim) %>%  # t transposes the matrix
  as_tibble() %>%
  mutate(SNP_ID=colnames(bim), cM=0) 

#head(bim) # now we have a table where each row is a SNP

bim <- bim %>% 
  separate(SNP_ID,into = c("chromosome","position"),sep="_",remove=FALSE) %>% # create a column for chromosome and position
  select(chromosome, SNP_ID, cM, position, allele1=V1, allele2=V2) # get columns in right order

write.table(bim,file="../output/rice.data.fastStructure.input.bim",col.names = FALSE, row.names = FALSE, quote = FALSE)
head(bim)
```

```{r Exercise 6 fastStructure}
#In the terminal, have to be in the output dir
#fastStructure -K 4 --input=rice.data.fastStructure.input --output=rice.fastStructure.out --format=str

fs_results <- read_delim("../output/rice.fastStructure.out.4.meanQ", delim="  ", col_names = FALSE, col_types = 'nnnn')
head(fs_results)

fs_results <- fs_results %>% 
  mutate(ID=data.geno.10000$ID) %>% 
  select(ID, pop1=X1, pop2=X2, pop3=X3, pop4=X4)
head(fs_results)

fs_results$assignedPop <- apply (fs_results[,-1], 1, which.max)
head(fs_results)

fs_results$maxPr <- apply(fs_results[,2:5],1,max) 
fs_results <- fs_results %>% 
  arrange(assignedPop,desc(maxPr)) %>%
  mutate(plot.order=row_number())
table(fs_results$assignedPop)
```

```{r Exercise 6 Reshaping data}
fs_results_long <- fs_results %>%
  pivot_longer(pop1:pop4, names_to = "population", values_to = "proportion")
               
head(fs_results_long)

fs_results_long %>%
  ggplot(aes(x=plot.order, y=proportion, color=population, fill=population)) + 
  geom_col()  +
  ylab("genome proportion") + 
  scale_color_brewer(type="div") + scale_fill_brewer(type="div")
```

```{r Exercise 6}

fs_results <- fs_results %>%
  mutate(assignedPop = as.character(assignedPop))
#Data exploration
#head(fs_results)
#head(PCs)
#str(fs_results)
#str(PCs)
geno.pca.pop <- left_join(fs_results, PCs, by = "ID")
#head(geno.pca.pop)

ggplot(data = geno.pca.pop, aes(x = PC1, y = PC2, color = assignedPop)) +
  geom_point() +
  ggtitle("PC1 and PC2 - Assigned Pop")

ggplot(data = geno.pca.pop, aes(x = PC2, y = PC3, color = assignedPop)) +
  geom_point() +
  ggtitle("PC2 and PC3 - Assigned Pop")

save(data.pheno, geno.pca, PCs, geno.pca.pop, fs_results, file = "../output/data_from_SNP_lab.Rdata")

```


- These populations relate to the PCA plots as the population assignments were based on the clusters of the PCA plots. The surprising part for the PC1 vs PC2 plot was that the cluster on the left that I thought was only one population was actually two populations. In the PC2 vs PC3 plot, the 4 different populations are better distinguished. 



**MV** -0

