# LAB: "Working with Illumina Sequence data, Viewing Illumina reads in IGV; finding SNPs, and Illumina Countinued: Analyze a VCF file in R"
## Working with Illumina Sequence Data Summary:
- Working with B.rapa illumina RNAseq data
- Goals:
  - Learn about Illumina reads, how to map them and QC
  - How to view reads in a genome browswer and how to find SNPs
  - Find genes that are differentailly expressed bw gt or trt
  - Build a GRN to det how genes conect to another
-Background:
  - Illumina sequencing
    - Raw outputs of illumina sequencer are `fastq` files
    - 350MM of 50-250 bp per lane
    - Multiplexing is used to process multiple samples at a time b/c 350M is a lot
      - Done by having some barcode to say where a fragment comes from
        - In line barcode added to the endds of the adapter adj to DNA of interest, need to trim off barcodes
        - Index tags are internal in the adapter which can be autoamatically sorted
  - QC
    - Illuminal sequencer assigned a PHRED score Q
      - Ranges from 0 to 41 and encoded in ASCII
      - Q = 40 is 1:10,000 mismatch, Q = 30 is 1:1,000 mismatch
  - Mapping
    - goals for this is to
      - find SNPs b/w the two gt's
      - Find DGE
      - To achieve goal need to know where reads map on genome
    - Multiple read mapping programs, `bwa` and `bowtie2`
    - We are mapping RNAseq data to a genomic reference so we need one to splice across introns
    - `tophat` uses `bowtie2` to map across introns
- Outline
  - Check FASTQ quality
  - Trim reads to keep high-quality reads
  - Split reads based on barcodes
  - Map to reference genome
- Need to download
  - `multiqc` to aggregate reports of QC from QC programs like fastQC
  - `trimmomatic` to filter reads based on certain criteria
  - `auto_barcode` to split fastq reads based on barcodes
- CHECK FASTQ QUALITY
  - use fastqc to import fastq file
    - compare results of good and bad sequences
- TRIM READS TO KEEP HIGH-QUALITY READS
  - our criteria for this project is rm when Q >= 20 over 4 bases and if the read is 50 or more bp loong
  - Can observe a difference in fastq with filtered reads
- BARCODE SPLITTING
  - use `auto_barcode` to split the reads out
  - rerun falast qc and look at the multiqc results
- MAPPING
  - use `tophat` to align sequence libraries to the reference genome
## Viewing Illumina reads in IGV; finding SNPs Summary:
- Goals:
  - learn about seq alignment and mapping
  - examine mapped reads in IGV
  - find polymorphic pos in seq data
- Looking a `tophat` outputs
  - accepted hits bam file for reads mapped
  - unmapped hits bam file for reads unmapped
  - indel bed showing insertions and deletions
  - jxns bed giving introns
  - align summary txt
- use `samtools` to look and manipulate bam/sam files
- use `IGV` to look at reads in context to reference
- CALLING SNPS
  - use `FreeBayes` to find polymorphisms b/w the two gt
  - rm dupicate DNA fragments using `sam tools`
  - create an index of the new files
  - use `freebayes` to look for SNPs
## Illumina Countinued: Analyze a VCF file in R
- `freebayes` output is a vcf file
  - vcf file contains info on polymorphisms, tsv
  - want to filter out lo Q SNPs
  - need to reformat once imported into R
- Look at quality and filter
  - filter using R, visualize quality dist via histogram
- Look at allele classes
  - filter so that only have SNPs that have 20+ depth
- Use IGV to look at different SNPs
  - see that 0/0, 0/1, 1/1 thing in vcf corresponds to homozR homozD, and hets
  - only 0/1 SNPs useful for F2 downstream mapping b/c different bw the two gt's
  - 
